bazel: add rp_util and compat

rockwotj opened the pull request:
Add the compat module to bazel, add rp_util to Bazel, fix schema printing tool.

Fixes: CORE-7462, CORE-8753


## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
bazel: add compat module

rockwotj committed:
bazel: add rp_util

rockwotj committed:
bazel: add compat/tests

Fixes: CORE-7462

rockwotj committed:
rputil: fix cluster config schema generation

At some point we switched to using a body_writer for this instead of
string, but that broke rputil, fix it by writing all the body to an
iobuf then printing it to stdout.

Fixes: CORE-8753

bharathv commented on src/v/rp_util/BUILD:
```diff
@@ -0,0 +1,18 @@
+
+redpanda_cc_binary(
+    name = "rp_util",
+    testonly = True,
```
I'm not 100% clear of the mplications of 'testonly=True', documentation says

```
This attribute is intended to mean that the target should not be contained in binaries that are released to production.
```

this particular binary is included in the final deb package IIRC.

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/24732/files#diff-d6738500e5c4a3fb8b7ff7e6f5c48e7617d4d310eca23adfe912b0b6993b65cdL48-L57

I looked at the packages recently and there is no rp_util.

In reality testonly=True means you can depend on test dependencies, and the compat package depends a lot random data generators that are marked as testonly, so it can must be marked testonly to depend on it.

I don't know how useful the json generators and checks are in the compat checks TBH, I think it requires manual effort to keep up to date and we don't really do that. Maybe C++ reflection will be our savior?

bharathv replied:
lol I added that comment 🙈 

you're right, I just built the pkgs locally and I don't see it, I confused it with something else.

> I don't know how useful the json generators and checks are in the compat checks TBH, I think it requires manual effort to keep up to date and we don't really do that. Maybe C++ reflection will be our savior?

I agree. Its a PITA to update it every time there is a change in topic properties. Can it be removed altogether @dotnwat ?

dotnwat replied:
it's all part of an effort to create automation around verification of data written to stable storage across releases. we aren't using it, but i guess we have to update it occasionally to keep it alive.

i think we can remove it no problem. easy enough to bring back if we ever want to restart that project.

bharathv reviewed (dismissed):
thanks for fixing the config json stuff.

rockwotj committed:
compat: fix clang-tidy warnings

The organization here of the headers was a mess, because the headers
only needed to work out at the end when included in `run.cc`. However,
bazel runs clang-tidy on header only libraries, so we actually need
each header to compile standalone.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--17c1ec9c-b190-4c64-90d3-42b6a883b419-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60469">build#60469</a></summary>


| test_id                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit                 | unit        | https://buildkite.com/redpanda/redpanda/builds/60469#01944959-d0a2-4bbf-86b6-45d576ef0022 | FLAKY         | 1/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/60469#01944959-d0a1-405e-9f59-bdbb37aba259 | FLAKY         | 1/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/60469#01944959-d0a2-4bbf-86b6-45d576ef0022 | FLAKY         | 1/2      |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60469#019449a0-a686-41b3-8e10-6c8c93fdfa88 | FLAKY         | 4/6      |
</details>
<!--/body-->

rockwotj commented:
/backport v24.3.x
